[java] Parsing error with try-with-resources and qualified resource

**Affects PMD Version:** 6.41.0
**Description:** PMD fails to parse the following file
**Exception Stacktrace:**
```xml
<error filename="/Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java" msg="PMDException: Error while parsing /Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java">
<![CDATA[net.sourceforge.pmd.PMDException: Error while parsing /Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:124)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:100)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:62)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:85)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Encountered " ")" ") "" at line 798, column 44.
Was expecting one of:
    "this" ...
    <IDENTIFIER> ...
    "@" ...
    <IDENTIFIER> ...
    
	at net.sourceforge.pmd.lang.java.ast.JavaParser.generateParseException(JavaParser.java:13532)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.jj_consume_token(JavaParser.java:13416)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableDeclaratorId(JavaParser.java:1974)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Resource(JavaParser.java:6876)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Resources(JavaParser.java:6809)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ResourceSpecification(JavaParser.java:6773)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TryStatement(JavaParser.java:6715)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5508)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5465)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.IfStatement(JavaParser.java:6278)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5481)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5465)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.IfStatement(JavaParser.java:6283)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5481)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclaration(JavaParser.java:2130)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBodyDeclaration(JavaParser.java:1811)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBody(JavaParser.java:1764)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceDeclaration(JavaParser.java:905)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TypeDeclaration(JavaParser.java:807)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.CompilationUnit(JavaParser.java:527)
	at net.sourceforge.pmd.lang.java.AbstractJavaParser.parse(AbstractJavaParser.java:62)
	at net.sourceforge.pmd.lang.AbstractParser.doParse(AbstractParser.java:44)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:136)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:200)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:118)
	... 10 more
]]>
</error>
```

**Code Sample demonstrating the issue:**
```java
import java.io.InputStream;
public class Foo {
    public InputStream in;
    public void bar() {
        Foo f = this;
       
        try (f.in) {
        }

        try (this.in) {
        }

    }
}
```

File: java.base/sun/security/ssl/SSLSocketImpl.java

Maybe this here? https://github.com/openjdk/jdk/blob/dfacda488bfbe2e11e8d607a6d08527710286982/src/java.base/share/classes/sun/security/ssl/SSLSocketImpl.java#L798

Looks like a paste error?

<error filename="/Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java" msg="PMDException: Error while parsing /Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java">
<![CDATA[net.sourceforge.pmd.PMDException: Error while parsing /Library/Java/JavaVirtualMachines/jdk-17.jdk/Contents/Home/lib/src.zip:java.base/sun/security/ssl/SSLSocketImpl.java
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:124)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:100)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:62)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:85)
	at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:29)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Encountered " ")" ") "" at line 798, column 44.
Was expecting one of:
    "this" ...
    <IDENTIFIER> ...
    "@" ...
    <IDENTIFIER> ...
    
	at net.sourceforge.pmd.lang.java.ast.JavaParser.generateParseException(JavaParser.java:13532)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.jj_consume_token(JavaParser.java:13416)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.VariableDeclaratorId(JavaParser.java:1974)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Resource(JavaParser.java:6876)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Resources(JavaParser.java:6809)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ResourceSpecification(JavaParser.java:6773)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TryStatement(JavaParser.java:6715)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5508)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5465)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.IfStatement(JavaParser.java:6278)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5481)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5465)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.IfStatement(JavaParser.java:6283)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Statement(JavaParser.java:5481)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.BlockStatement(JavaParser.java:5671)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.Block(JavaParser.java:5588)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclaration(JavaParser.java:2130)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBodyDeclaration(JavaParser.java:1811)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBody(JavaParser.java:1764)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceDeclaration(JavaParser.java:905)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.TypeDeclaration(JavaParser.java:807)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.CompilationUnit(JavaParser.java:527)
	at net.sourceforge.pmd.lang.java.AbstractJavaParser.parse(AbstractJavaParser.java:62)
	at net.sourceforge.pmd.lang.AbstractParser.doParse(AbstractParser.java:44)
	at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:136)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:200)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:118)
	... 10 more
]]>
</error>

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -16,6 +16,9 @@ This is a {{ site.pmd.release_type }} release.
 
 ### Fixed Issues
 
+*   java
+    *   [#3698](https://github.com/pmd/pmd/issues/3697): \[java] Parsing error with try-with-resources and qualified resource
+
 ### API Changes
 
 ### External Contributions

--- a/pmd-java/etc/grammar/Java.jjt
+++ b/pmd-java/etc/grammar/Java.jjt
@@ -2406,7 +2406,7 @@ void Resources() :
 void Resource() :
 {}
 {
-   LOOKAHEAD("this" | Name() ")") (
+   LOOKAHEAD("this" | Name() (")" | ";")) (
        {checkForBadConciseTryWithResourcesUsage();}
        Name()
        // replaced with Expression in PMD 7, do the bare minimum

--- a/pmd-java/etc/grammar/Java.jjt
+++ b/pmd-java/etc/grammar/Java.jjt
@@ -2406,9 +2406,15 @@ void Resources() :
 void Resource() :
 {}
 {
-   LOOKAHEAD(2) ( ( "final" {jjtThis.setFinal(true);} | Annotation() )* LocalVariableType() VariableDeclaratorId() "=" Expression() )
-   |
-   Name() {checkForBadConciseTryWithResourcesUsage();}
+   LOOKAHEAD("this" | Name() ")") (
+       {checkForBadConciseTryWithResourcesUsage();}
+       Name()
+       // replaced with Expression in PMD 7, do the bare minimum
+       | "this" "." Name() // possible pmd6 improvement: add a isThisModifier() or so
+   )
+   | ( "final" {jjtThis.setFinal(true);} | Annotation() )*
+     LocalVariableType() VariableDeclaratorId() "=" Expression()
+
 }
 
 void CatchStatement() :

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ParserCornersTest.java
@@ -25,6 +25,7 @@ public class ParserCornersTest {
     private static final String MULTICATCH = "public class Foo { public void bar() { "
         + "try { System.out.println(); } catch (RuntimeException | IOException e) {} } }";
     private final JavaParsingHelper java = JavaParsingHelper.WITH_PROCESSING.withResourceContext(ParserCornersTest.class);
+    private final JavaParsingHelper java9 = java.withDefaultVersion("9");
     private final JavaParsingHelper java8 = java.withDefaultVersion("1.8");
     private final JavaParsingHelper java4 = java.withDefaultVersion("1.4");
     private final JavaParsingHelper java5 = java.withDefaultVersion("1.5");
@@ -88,6 +89,33 @@ public class ParserCornersTest {
     @Test
     public final void testCastLookaheadProblem() {
         java4.parse(CAST_LOOKAHEAD_PROBLEM);
+    }
+
+    @Test
+    public final void testTryWithResourcesConcise() {
+        // https://github.com/pmd/pmd/issues/3697
+        java9.parse("import java.io.InputStream;\n"
+                        + "public class Foo {\n"
+                        + "    public InputStream in;\n"
+                        + "    public void bar() {\n"
+                        + "        Foo f = this;\n"
+                        + "        try (f.in) {\n"
+                        + "        }\n"
+                        + "    }\n"
+                        + "}");
+    }
+
+    @Test
+    public final void testTryWithResourcesThis() {
+        // https://github.com/pmd/pmd/issues/3697
+        java9.parse("import java.io.InputStream;\n"
+                        + "public class Foo {\n"
+                        + "    public InputStream in;\n"
+                        + "    public void bar() {\n"
+                        + "        try (this.in) {\n"
+                        + "        }\n"
+                        + "    }\n"
+                        + "}");
     }
 
     /**

--- a/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/jdk9_try_with_resources.java
+++ b/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/jdk9_try_with_resources.java
@@ -2,6 +2,13 @@ public class InputJava9TryWithResources {
  public static void main() {
   MyResource resource1 = new MyResource();
   MyResource resource2 = new MyResource();
+  try (resource1) { }
+  try (resource1;) { }
   try (resource1; resource2) { }
+  try (resource1.foo) { }
+  try (resource1.foo.a) { }
+  try (resource1.foo.Type v = null) { }
+  try (this.foo.aa) { }
+  try (this.foo) { }
  }
 }

